public/app/features/explore/QueryRows.tsx

```diff
@@ -84,22 +92,47 @@ export const QueryRows = ({ exploreId }: Props) => {
       onQueryCopied={onQueryCopied}
       onQueryRemoved={onQueryRemoved}
       onQueryToggled={onQueryToggled}
+      onBeginQuerySave={onBeginQLSave}
       data={queryResponse}
       app={CoreApp.Explore}
       history={history}
       eventBus={eventBridge}
-      queryRowWrapper={(children, refId) => (
-        <ContentOutlineItem
-          title={refId}
-          icon="arrow"
-          key={refId}
-          panelId="Queries"
-          customTopOffset={-10}
-          level="child"
-        >
-          {children}
-        </ContentOutlineItem>
-      )}
+      queryRowWrapper={(children, refId) => {
+        const query = queries.find((q) => q.refId === refId);
+        return (
+          <>
+            <ContentOutlineItem
+              title={refId}
+              icon="arrow"
+              key={refId}
+              panelId="Queries"
+              customTopOffset={-10}
+              level="child"
+            >
+              {children}
+            </ContentOutlineItem>
+            {query !== undefined && (
+              <Modal
```

Comment:
This does add one modal per query, but the previous method had one modal per "add to library" button. Should this simply be in Explore and be controled via state?

Reply:
If I understand it right now, the button to add to QL is inside the `QueryEditorRows` component and it controls state of the parent via a prop function that then triggers the modal here, right? Is there a benefit in that? Making the modal part of the QL save button seems like it would package this more nicely, so that any need to move it around would be easier, and also you would see all the logic in one place. Is there any need for the `QueryRows` component to know about the QL save button?

Reply:
Nope, I was just wondering if we would have a preference to do it another way - seems how I have it is good :D

Reply:
Actually with some changes in how we need to initiate the extra query action, it made sense to move everything up to the main Explore page. We still don't have extra stuff in the app state though, we can control it all in the component state.